radixdlt/src/main/java/com/radixdlt/consensus/BFTEventPreprocessor.java

```diff
@@ -0,0 +1,221 @@
+import com.radixdlt.consensus.liveness.ProposerElection;
+import com.radixdlt.counters.SystemCounters;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.crypto.Hash;
+import java.util.Objects;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Preprocesses consensus events and ensures that the vertexStore is synced to
+ * the correct state before they get forwarded to the actual state reducer.
+ *
+ * This class should not be updating any part of the BFT Safety state besides
+ * the VertexStore.
+ *
+ * A lot of the queue logic could be done more "cleanly" and functionally using
+ * lambdas and Functions but the performance impact is too great.
+ *
+ * This class is NOT thread-safe.
+ */
+public final class BFTEventPreprocessor implements BFTEventProcessor {
+	private static final Logger log = LogManager.getLogger();
+
+	private final ECPublicKey myKey;
+	private final BFTEventProcessor forwardTo;
+	private final VertexStore vertexStore;
+	private final PacemakerState pacemakerState;
+	private final ProposerElection proposerElection;
+	private final SystemCounters counters;
+	private final SyncQueues queues;
+
+	public BFTEventPreprocessor(
+		ECPublicKey myKey,
+		BFTEventProcessor forwardTo,
+		PacemakerState pacemakerState,
+		VertexStore vertexStore,
+		ProposerElection proposerElection,
+		SyncQueues queues,
+		SystemCounters counters
+	) {
+		this.myKey = Objects.requireNonNull(myKey);
+		this.pacemakerState = Objects.requireNonNull(pacemakerState);
+		this.vertexStore = Objects.requireNonNull(vertexStore);
+		this.proposerElection = Objects.requireNonNull(proposerElection);
+		this.queues = queues;
+		this.counters = Objects.requireNonNull(counters);
+		this.forwardTo = forwardTo;
+	}
+
+	private String getShortName() {
+		return myKey.euid().toString().substring(0, 6);
+	}
+
+	private boolean peekAndExecute(SyncQueue queue, Hash vertexId) {
+		final RequiresSyncConsensusEvent event = queue.peek(vertexId);
+		if (event == null) {
+			return false;
+		}
+
+		// Explicitly using switch case method here rather than functional method
+		// to process these events due to much better performance
+		if (event instanceof NewView) {
+			final NewView newView = (NewView) event;
+			return this.processNewViewInternal(newView);
+		}
+
+		if (event instanceof Proposal) {
+			final Proposal proposal = (Proposal) event;
+			return this.processProposalInternal(proposal);
+		}
+
+		throw new IllegalStateException("Unexpected consensus event: " + event);
+	}
+
+	/**
+	 * Signal that vertexStore now contains the given vertexId.
+	 * Executes events which are dependent on this vertex
+	 *
+	 * @param vertexId the id of the vertex which is now guaranteed be synced.
+	 */
+	public void processLocalSync(Hash vertexId) {
+		for (SyncQueue queue : queues.getQueues()) {
+			boolean first = true;
+			while (peekAndExecute(queue, first ? vertexId : null)) {
+				queue.pop();
+				first = false;
+			}
```

Comment:
Might be more performant to have:
```
Hash firstVertexId = vertexId;
while (peekAndExecute(queue, firstVertexId)) {
    queue.pop();
    firstVertexId = null;
}
```
Then again, maybe the fancy JITs do this automagically.

Reply:
or even more performant:
```
if (peekAndExecute(queue, vertexId)) {
    queue.pop();
    while (peekAndExecute(queue, null)) {
        queue.pop();
    }
}
```